src/kernel/kernel.h

```diff
@@ -55,7 +55,11 @@ void init_kernel_context(kernel_context kc, int type, int size, queue free_ctx_q
 #define cpu_interrupt 3
 #define cpu_user 4
 
-extern boolean shutting_down;
+#define SHUTDOWN_POWER (1<<0) // shutdown triggered externally
+#define SHUTDOWN_TERM  (1<<1) // process termination triggered by the kernel as a result of a signal
+#define SHUTDOWN_EXIT  (1<<2) // process termination triggered by the process itself
```

Comment:
Thinking more about it, we can do without the SHUTDOWN_TERM and SHUTDOWN_EXIT flags (in the future, perhaps we could incorporate this info in the kernel_shutdown() arguments); instead, we need a SHUTDOWN_ONGOING flag, which will be set in kernel_shutdown(), and will be checked instead of checking the existing boolean variable.
So all the callers to kernel_shutdown() don't need to touch the `shutting_down` variable.